Applications/Spire/Source/Ui/TextAreaBox.cpp

```diff
@@ -0,0 +1,634 @@
+    &TextAreaBox::on_text_changed);
+  m_placeholder = new ElidedLabel(this);
+  m_placeholder->setFixedSize(0, 0);
+  m_stacked_widget->addWidget(m_placeholder);
+  m_placeholder->setCursor(m_text_edit->cursor());
+  m_placeholder->setAttribute(Qt::WA_TransparentForMouseEvents);
+  m_scroll_box = new ScrollBox(m_stacked_widget, this);
+  m_scroll_box->set(ScrollBox::DisplayPolicy::NEVER,
+    ScrollBox::DisplayPolicy::ON_OVERFLOW);
+  m_scroll_box->get_vertical_scroll_bar().adjustSize();
+  m_scroll_box->setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Minimum);
+  m_scroll_box->installEventFilter(this);
+  m_scroll_box->findChild<ScrollableLayer*>()->installEventFilter(this);
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  layout->addWidget(m_scroll_box);
+  proxy_style(*this, *m_scroll_box);
+  add_pseudo_element(*this, Placeholder());
+  connect_style_signal(*this, [=] { on_style(); });
+  connect_style_signal(*this, Placeholder(), [=] { on_style(); });
+  set_style(*this, DEFAULT_STYLE());
+  connect(m_text_edit, &QTextEdit::cursorPositionChanged, this,
+    &TextAreaBox::on_cursor_position);
+  m_current_connection = m_model->connect_current_signal(
+    [=] (const auto& value) { on_current(value); });
+  connect(m_text_edit->document()->documentLayout(),
+    &QAbstractTextDocumentLayout::documentSizeChanged,
+    [=] (const auto& size) { m_stacked_widget->adjustSize(); });
+}
+
+const std::shared_ptr<TextModel>& TextAreaBox::get_model() const {
+  return m_model;
+}
+
+const QString& TextAreaBox::get_submission() const {
+  return m_submission;
+}
+
+void TextAreaBox::set_placeholder(const QString& value) {
+  m_placeholder_text = value;
+  update_placeholder_text();
+}
+
+bool TextAreaBox::is_read_only() const {
+  return m_text_edit->isReadOnly();
+}
+
+void TextAreaBox::set_read_only(bool read_only) {
+  m_text_edit->setReadOnly(read_only);
+  m_scroll_box->get_vertical_scroll_bar().set_position(0);
+  if(read_only) {
+    m_scroll_box->set_vertical(ScrollBox::DisplayPolicy::NEVER);
+    match(*this, ReadOnly());
+  } else {
+    m_scroll_box->set_vertical(ScrollBox::DisplayPolicy::ON_OVERFLOW);
+    unmatch(*this, ReadOnly());
+  }
+  update_layout();
+}
+
+connection TextAreaBox::connect_submit_signal(
+    const SubmitSignal::slot_type& slot) const {
+  return m_submit_signal.connect(slot);
+}
+
+QSize TextAreaBox::sizeHint() const {
+  return m_scroll_box->sizeHint();
+}
+
+void TextAreaBox::changeEvent(QEvent* event) {
+  if(event->type() == QEvent::EnabledChange) {
+    update_placeholder_text();
+  }
+  QWidget::changeEvent(event);
+}
+
+bool TextAreaBox::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_text_edit && event->type() == QEvent::FocusOut) {
+    m_submission = m_model->get_current();
+    m_submit_signal(m_submission);
+  } else if(event->type() == QEvent::Wheel) {
+    if(m_text_edit->isReadOnly()) {
```

Comment:
Why is wheeling disabled on read-only?

Reply:
I don't adjust the size of the text edit when it's read-only so it's possible to scroll it. I can probably clean this up and remove the call to findChild in the above comment by resizing the text edit when it's read-only.

Reply:
Good, clean it up.